src/Platform/Sdl/SdlPlatformSharedState.cpp

```diff
@@ -47,49 +47,54 @@ SdlWindow *SdlPlatformSharedState::window(uint32_t id) const {
     return ValueOr(_windowById, id, nullptr);
 }
 
-void SdlPlatformSharedState::registerGamepad(SdlGamepad *gamepad) {
-    int32_t id = nextFreeGamepadId();
-    assert(id >= 0);
+void SdlPlatformSharedState::initializeGamepads() {
+    for (int i = 0; i < SDL_NumJoysticks(); i++) {
+        if (!SDL_IsGameController(i))
+            continue;
 
-    _gamepadById[id] = gamepad;
+        initializeGamepad(i);
+    }
 }
 
-void SdlPlatformSharedState::unregisterGamepad(SdlGamepad *gamepad) {
-    for (auto it = _gamepadById.begin(); it != _gamepadById.end(); it++) {
-        if (gamepad == it->second) {
-            _gamepadById.erase(it);
-            return;
-        }
-    }
+SdlGamepad *SdlPlatformSharedState::initializeGamepad(int gamepadId) {
+    assert(SDL_IsGameController(gamepadId));
 
-    assert(0); //shouldn't happen
-}
+    std::unique_ptr<SDL_GameController, void(*)(SDL_GameController *)> gamepad(SDL_GameControllerOpen(gamepadId), &SDL_GameControllerClose);
+    if (!gamepad) {
+        logSdlError("SDL_GameControllerOpen");
+        return nullptr;
+    }
 
-int32_t SdlPlatformSharedState::getGamepadIdBySdlId(uint32_t id) {
-    int32_t ret = -1;
+    SDL_Joystick *joystick = SDL_GameControllerGetJoystick(gamepad.get());
+    if (!joystick) {
+        logSdlError("SDL_GameControllerGetJoystick");
+        return nullptr;
+    }
 
-    for (auto it = _gamepadById.begin(); it != _gamepadById.end(); it++) {
-        SdlGamepad *gamepad = it->second;
-        if (gamepad->gamepadSdlId() == id)
-            return gamepad->id();
+    SDL_JoystickID id = SDL_JoystickInstanceID(joystick);
+    if (id < 0) {
+        logSdlError("SDL_JoystickInstanceID");
+        return nullptr;
     }
+    assert(!_gamepadById.contains(id)); // However, the code below will work fine even if id is there.
```

Comment:
This assert triggers when using my xbox360 controller. Commented out and works as expected.